File: ML-Practicas/P2 - Linear Regresion/polinomial_regression.py
# -*- coding: utf-8 -*-
"""
Created on Fri Jun 21 16:09:19 2019

@author: adrian.perez
"""

#Regresión polinomica

# Plantilla de Preprocesado 
#Como importar las librerias
import numpy as np
import matplotlib.pyplot as plt
import pandas as pd

#importar el dataset
dataset = pd.read_csv("Position_salaries.csv")

X=dataset.iloc[:,1:2].values #EStablece las variables independientes
y=dataset.iloc[:, 2].values #establecer variable dependiente

#Dividir el dataset en conjuto de entrenamiento y conjunto de test
#En este caso no se dividen en sets  de test y training por que hay pocos datos

#from sklearn.model_selection import train_test_split
#X_train,X_test,y_train,y_test=train_test_split(X,y,test_size=1/3,random_state=0)

#Escalado de vaiables
"""from sklearn.preprocessing import StandardScaler
sc_X=StandardScaler()
X_train=sc_X.fit_transform(X_train)
X_test=sc_X.transform(X_test)"""
#Ajustar la regresion lineal con el dataset
from sklearn.linear_model import LinearRegression
lineal_reg=LinearRegression()
lineal_reg.fit(X,y)

# Ajustar regresión polinomica con el data set
from sklearn.preprocessing import PolynomialFeatures
poly_reg=PolynomialFeatures(degree=5)
X_poly=poly_reg.fit_transform(X)
lin_reg_2=LinearRegression()
lin_reg_2.fit(X_poly,y)

#Visualizacion de los resultados del modelo lineal

plt.scatter(X,y,color="red")
plt.plot(X,lineal_reg.predict(X),color="blue")
plt.title("Modelo de regresion lineal")
plt.xlabel("Posición del empleado")
plt.ylabel("Sueldo en $")
plt.show()

#visualizacipon de los resultados del  modelo polinómico
X_grid=np.arange(min(X),max(X),0.1) # Ajustar los valores de X para un mejro calidad en la grafica(unicamente)
X_grid=X_grid.reshape(len(X_grid),1)
plt.scatter(X,y,color="red")
#plt.plot(X,lin_reg_2.predict(X_poly),color="blue")
plt.plot(X_grid,lin_reg_2.predict(poly_reg.fit_transform(X_grid)),color="blue")
plt.title("Modelo de regresion polinomica")
plt.xlabel("Posición del empleado")
plt.ylabel("Sueldo en $")
plt.show()


#Predicción de los modelos

lineal_reg.predict([[6.5]])

lin_reg_2.predict(poly_reg.fit_transform([[6.5]]))

File: ML-Practicas/P2 - Linear Regresion/polinomial_regression.R
# Regresón Polinómica

dataset=read.csv('Position_salaries.csv')
dataset=dataset[,2:3] #Solo ocupamos las columnas 2 y 3

#dividir los datos en conjuntos de entrenamiento y test
#install.packages("caTools")
library(caTools)
set.seed(123) #Semilla para aleatorios

#Dividir en data set de entrenamiento y test
#En este caso no se dividen porque hay pocos datos

# split=sample.split(dataset$Salary,SplitRatio =2/3)
# training_set=subset(dataset,split==TRUE)
# testing_set=subset(dataset,split==FALSE)

#Tampoco hay proceso de escalado

#Ajustar Modelo re Regresión Lineal con el Conunto de datos

lin_reg=lm(formula = Salary ~.,
           data = dataset)

#Ajustar Modelo de Regresión Polinómica con el Conjunto de datos
dataset$Level2=dataset$Level^2 #Creamos columna de los cuadrados de la variable independiente
dataset$Level3=dataset$Level^3 #Creamos columna con el cubo de la variable independiente
dataset$Level5=dataset$Level^5 
poly_reg=lm(formula = Salary ~.,
            data = dataset)

#Visualización del modelo lineal
#install.packages("ggplot2")
#library(ggplot2)
ggplot()+
  geom_point(aes(x=dataset$Level,y=dataset$Salary),
             color="red")+
  geom_line(aes(x=dataset$Level,y=predict(lin_reg,newdata = dataset)),
            color="blue")+
  ggtitle("Predicción lineal del sueldo en función del nivel del empleado")+
  xlab("Nivel del empleado")+
  ylab("Sueldo $")


#Visualización del modelo polinómico
dataset$predict=predict(poly_reg,newdata = dataset)

ggplot()+
  geom_point(aes(x=dataset$Level,y=dataset$Salary),
             color="red")+
  geom_line(aes(x=dataset$Level,y=predict(poly_reg,newdata = dataset)),
            color="blue")+
  ggtitle("Predicción polinómica del sueldo en función del nivel del empleado")+
  xlab("Nivel del empleado")+
  ylab("Sueldo $") + 
  geom_text(aes(x=dataset$Level,y=predict(poly_reg,newdata = dataset),label=dataset$predict),hjust=0, vjust=0)

#Predicción de resultados con Regresión Lineal
y_pred=predict(lin_reg,newdata = data.frame(Level=6.5))

#Predicción de resultados con Regresión Polinómica
y_pred_poly=predict(poly_reg,newdata = data.frame(Level=6.5,
                                             Level2=6.5^2,
                                             Level3=6.5^3,
                                             Level5=6.5^5))
